### ISSUE TITLE
Cannot use proxy to connect to hipchat.

### ISSUE THREAD

**Issue by shadow7412** [skoczen/will#147]
Looks like the proxy settings aren't being set in all the right places?
```
Traceback (most recent call last):
  File "/home/steven/ENV/local/lib/python2.7/site-packages/will/main.py", line 367, in bootstrap_xmpp
    self.connect()
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/clientxmpp.py", line 160, in connect
    reattempt=reattempt)
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 448, in connect
    args=(reattempt,))
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/thirdparty/statemachine.py", line 70, in transition
    func=func, args=args, kwargs=kwargs)
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/thirdparty/statemachine.py", line 111, in transition_any
    return_val = func(*args,**kwargs) if func is not None else True
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 508, in _connect
    connected = self._connect_proxy()
  File "/home/steven/ENV/local/lib/python2.7/site-packages/sleekxmpp/xmlstream/xmlstream.py", line 599, in _connect_proxy
    password = self.proxy_config['password']
KeyError: u'password'
```

**Comment by skoczen** [comment]
Hey @shadow7412,

Thanks for this report!  I'm pretty sure this is the same bug that's fixed in #130.  Would you be able to test that out by pip installing that fork?

I'm hoping to get it merged in with a sprint toward the end of this week.

Thanks for your patience!

**Comment by shadow7412** [comment]
Hmm. That didn't end well.
![image](https://cloud.githubusercontent.com/assets/286341/7825683/4da01802-0453-11e5-9812-347c1f5468fc.png)


I altered settings.py to `help(config)` to see what was actually in that module - and it looks right. So I have no idea why it's claiming that PLUGINS is undefined just a few lines later...

**Comment by skoczen** [comment]
Obvious question, but do you still get the same error without any alterations?

**Comment by shadow7412** [comment]
Correct - I made the modifications after I got the error originally.

**Comment by skoczen** [comment]
Very strange.   Usually, things like this mean that there's another `settings` module on your pythonpath before will's - or something equally strange.

In the interest of just getting you running, the other fix mentioned in that issue is to set `WILL_PROXY_URL` in the environment.  Want to give that a go, with normal trunk will?

**Comment by shadow7412** [comment]
Back on normal trunk now, tried it and got the original error.
Removed all proxy settings from config.py - still the same :(

**Comment by skoczen** [comment]
You're in a virtualenv, yeah?

**Comment by shadow7412** [comment]
Yep.

**Comment by shadow7412** [comment]
If it makes any difference, I was running will like this;
    WILL_PROXY_URL="http://username:password@hostname:8080" ./run_will.py

**Comment by skoczen** [comment]
It shouldn't, though you could `export` it first, and then `python run will` to be sure.

Or, more explicitly, call the `python` from your virtualenv.


Darn.  Ok, so let's think, here.  

Big picture, will is used by thousands of people, and I haven't seen this problem before.  This suggest (but certainly doesn't guarantee) that there's something different in your setup.

Can you give some debug info?

- `which python`
- `python -V`
- OS, version, etc
- is this local, or remote on a server?
- bash/tcsh/zsh?


Also, if, just before line 58 in `__main__`, you do a `print settings.__dict__` or `print dir(settings)`, what do you get?

Thanks for your help and patience in debugging this!

**Comment by shadow7412** [comment]
You're the one who should be thanked for their patience - and for developing such a great bot.

* which python
** /home/steven/ENV/bin/python
* python -V
** Python 2.7.6
* OS, version, etc
** Ubuntu 14.04.2 LTS
* Remote, sortof. It's in a Virtual Machine on the same computer.
* bash/tcsh/zsh?
** bash

As for the print, I did it and I can confirm the settings are there, including the password. I'm hesitant to post it due to my likely missing out on details that should be suppressed. Happy to take my word? Anything I should look for specifically?

**Comment by skoczen** [comment]
Many thanks for that - I'm actually as proud of the community we've developed here as much as the bot itself.  One of our big principles is to [be good to people](http://skoczen.github.io/will/improve/#making-will-better-and-better), especially folks who are taking time out of their day to troubleshoot.   In the long run, this also means that many happy people stick around and keep improving will.   At this point, I think the majority of the code in will is still mine, but it's getting pretty close to tipping over to the community.

TL;DR: You're very welcome - it's what we do here, and I'm glad it's working for you.

Bummed on your answers - they all look really normal.

All good on not posting sensitive info.  Most critically, do you see `PLUGINS` in there?

**Comment by shadow7412** [comment]
Sadly, yes. :crying_cat_face:

**Comment by skoczen** [comment]
Curiouser and curiouser.

Definite fighting-in-the-dark debugging, here, but, can you add, just before line 58,

```python
print settings.getattr('PLUGINS', None)
```

**Comment by shadow7412** [comment]
That'd be `print getattr(settings, "PLUGINS", None)` I assume.

`['will.plugins.admin', 'will.plugins.chat_room', 'will.plugins.devops', 'will.      plugins.friendly', 'will.plugins.help', 'will.plugins.productivity', 'will.plugins.web',       'plugins']`

But this is in the trunk branch that didn't have that problem - did you want to switch back to the other one?

**Comment by shadow7412** [comment]
(pressed the wrong comment button, sorry)

**Comment by skoczen** [comment]
Oh, goodness. I totally misunderstood - I thought you were still seeing the `PLUGINS` problem on the trunk.  Got it, now.

Ignore my last lines of exploration, then.  My brain's pretty mush-piled here and I'll just waste your time this evening.  Will pick this up tomorrow when I'm fresh!

**Comment by shadow7412** [comment]
Haha no problems :)

**Comment by skoczen** [comment]
Alright, with a fresh brain in the morning, I'm going to do the intelligent thing, and pull in the person who knows the most about the proxy code.  @dmuntean , does this error (the very top one) look at all familiar to you?  Thoughts?

**Comment by skoczen** [comment]
@shadow7412 Looking again, I'm still of the mind that this is the ordering problem.

Can you try making the changes in this commit locally (isolating any pip install weirdnesses), and paste the output?

https://github.com/dmuntean/will/commit/df68a7f2a7b9b4e334d803dadd6187cfc61530b1

**Comment by dmuntean** [comment]
@shadow7412 the initial error you are getting is because of a typo in proxy_config dict

@skoczen I added a commit to fix the typo, it should all be ok when it is merged

The second error is related to virtual env not being clean, it probably has leftovers from bot master and pull request branch. @shadow7412 can you please try creating a new clean virtual env and install pull request #130 now - I believe it should all be working ok now.

**Comment by skoczen** [comment]
@dmuntean fantastic, thanks for hopping on this so quickly!

@shadow7412 if you've got the time, it'd be great if you can try that out in the next day or so.  Unless anything catches fire, I should have an open-source block on Friday where I can get all these PRs merged.

Thanks!

**Comment by shadow7412** [comment]
A fresh virtualenv, and a s/sss/ss/, and we're back to settings not having PLUGINS defined.

I ran `help(config)` just after it got imported in settings.py, and it's the right one with the right settings.
There was also a `help(settings)` just before attempting to access settings.PLUGINS, which also showed the expected file in the correct virtualenv - but dir shows `['__builtins__', '__doc__', '__file__', '__name__', '__package__', 'columns', 'error', 'import_settings', 'indent', 'note', 'os', 'puts', 'show_valid', 'urlparse', 'warn']` - which clearly means that `import_settings` didn't really import the settings...

**Comment by skoczen** [comment]
And, that it's grabbing the right module.  Hm.

Now this really bugs me.  Give me a few minutes.

**Comment by skoczen** [comment]
Ok, finally reproduced, and found it.  Lines 153-161 are under-indented.  @dmuntean can you make the change in your PR?

@shadow7412 I reproduced your error exactly, and then it went away with that change.

My full settings.py is [here](https://gist.github.com/skoczen/eb1fd86408b5fad70027), if you'd rather try out with copy/paste.

Thanks again for all your patience!

**Comment by dmuntean** [comment]
@skoczen done via f6c648ee504a320b71293b0b8af59ca97b061b60

**Comment by shadow7412** [comment]
Fix confirmed - good spotting :)

**Comment by skoczen** [comment]
Fantastic, thanks @shadow7412 and @dmuntean for your efforts in tracking down and fixing this.  Hoping it'll go out tomorrow.
